Cloudupload nextcloud

DavHau opened the pull request:
Hey, this is my first PR here.
This adds cloud upload options for NextCloud and for other WebDav servers in general.
I developed and tested the following using docker:
- [x] Tesing uplaod service via 'Test Service' GUI button
- [x] Create missing directories automatically
- [x] Upload pictures and videos

While testing with docker i had the problem that i didn't have a camera connected. I wrote myself some dummy mjpeg network camera which streams cute kittens and changes their colors from time to time to simulate motion. It worked quite well for testing motioneye. Is there any better tool to use for developing? Otherwise i can add this also with another PR.

DavHau commented:
Now the arm pipeline takes 31 min because lxml pip package has to be compiled. Hope this is ok.

kleini commented:
> While testing with docker i had the problem that i didn't have a camera connected. I wrote myself some dummy mjpeg network camera which streams cute kittens and changes their colors from time to time to simulate motion. It worked quite well for testing motioneye. Is there any better tool to use for developing? Otherwise i can add this also with another PR.

Please add your camera simulation in another PR. This will be a great help for testing.

kleini commented on .travis.yml:
```diff
@@ -15,7 +15,7 @@ env:
   - docker info
   - docker run --rm --privileged multiarch/qemu-user-static:register --reset
-  - docker build --build-arg VCS_REF=$TRAVIS_COMMIT --build-arg BUILD_DATE=$(date +"%Y-%m-%dT%H:%M:%SZ") -t $DOCKER_REPO:$TRAVIS_BRANCH-$TARGET -f extra/Dockerfile${EXT} .
+  - travis_wait 40 docker build --build-arg VCS_REF=$TRAVIS_COMMIT --build-arg BUILD_DATE=$(date +"%Y-%m-%dT%H:%M:%SZ") -t $DOCKER_REPO:$TRAVIS_BRANCH-$TARGET -f extra/Dockerfile${EXT} .
```
Why is `travis_wait 40` needed?

DavHau replied:
I had to add this since compiling lxml for arm takes longer than 20 minutes. Travis times out after 10. Qemu emulation is making this so slow i guess.

kleini commented on extra/Dockerfile.armv7-armhf:
```diff
@@ -36,7 +36,11 @@ RUN apt-get update && \
     v4l-utils \
-    zlib1g-dev && \
+    zlib1g-dev \
+    build-essential \
```
Please keep the contents in Docker images minimal. Docker images are meant for runtime not for development. Feel free to install additional tools, when starting your container.

kleini replied:
Found it. `pip install lxml` needs the additional time. But where is it installed? I don't see commands for that. Am I blind?

DavHau replied:
lxml is a dependency of webdavclient3. webdavclient3 is defined as a requirement in setup.py.
The apt packages which I added to the arm Dockerfile are actually needed to successfully compile lxml. The ubuntu Dockerfile seems to not have this issue.

DavHau replied:
I added a pre stage docker container to the arm dockerfile which is only for building the modules. Later only the finished python wheels are copied to the production container. That removes the necessity of installing packages only needed for building inside the production docker container. Please have a look. After all the new arm container grows with this PR only 1MB (from 375M to 376M)

kleini commented on extra/Dockerfile.armv7-armhf:
```diff
@@ -50,7 +54,11 @@ RUN apt-get update && \
     python-pip \
     python-setuptools \
-    python-wheel && \
+    python-wheel \
```
Please keep the contents in Docker images minimal. Docker images are meant for runtime not for development. Feel free to install additional tools, when starting your container.

DavHau replied:
The apt packages which I added to the arm Dockerfile are actually needed to successfully compile lxml. The ubuntu Dockerfile seems to not have this issue.

DavHau replied:
All of the added dependencies are only needed for compiling and are uninstalled afterwards (see further in the Dockerfile). Except libxslt1.1 . It must be kept, otherwise the program crashes.

DavHau replied:
I now removed all the added apt dependencies into a pre staging docker container.

DavHau replied:
Also i removed zlib1g-dev which was not needed anymore already since you changed to install pillow from apt instead of pip.

bob-lee commented on motioneye/uploadservices.py:
```diff
@@ -14,28 +14,33 @@
 
 
-class UploadService(object):
+class UploadService:
```
Needs to test other services as UploadService and GoogleBase has changed.

DavHau replied:
What do you mean by test? Unit test, or just click testing through the UI?

bob-lee replied:
I mean regression test. As base classes that existing services are based on have changed, they (existing services) need to be tested as well to make sure there was no unwanted side effect on them.

bob-lee commented on setup.py:
```diff
@@ -63,7 +63,7 @@ def apply_patches(self, base_dir):
     packages=['motioneye'],
 
-    install_requires=['tornado>=3.1,<6', 'jinja2', 'pillow', 'pycurl'],
+    install_requires=['tornado>=3.1,<6', 'jinja2', 'pillow', 'pycurl', 'webdavclient3==0.12'],
```
Curious if webdav provides REST api so as to avoid adding webdavclient here?

DavHau replied:
Actually I used only 3 or so out of many webdav calls. Maybe the whole functionality could also be done by building the requests manually with pycurl or requests. Should i take a look into this? Still some answers of the webdav server might be in XML which has to be parsed. Also the webdav3 library handles errors nicely and throws meaningful exceptions.

bob-lee replied:
If a new upload service is doable without adding additional library, I would think that's the way to go to keep this project lightweight.

ccrisan replied:
I certainly don't want motionEye to pull in a dep like `webdavclient`. This should be optional.

DavHau commented:
> In main.html, wonder why ‘http|https’ and ‘ftp|sftp’ were removed in two places? Code changes here should not affect other upload services anyway.

First of all http/https should be removed from main.html completely in my opinion since it is nowhere used and there is no backend for it. I saw in the commit history that the http/https flags have been added with the first commits that added upload services. Probably a draft, that has never been implemented.
I think those should better be removed until there is actual functionality for it. But the decision is up to you guys.
Anyway, there only was a 'Server Address' input field before my PR and the description says clearly that it takes an IP address or a domain name. A URL input is needed in my case which contains a protocol, a domain name, a port and a path all in once, so i didn't want to reuse the 'Server Address' field since its description doesn't match and i added a 'Server URL' field.
Also http and https (if they were implemented) would take a URL and not a domain, so i unlinked them from 'Server Address' which is now only used for ftp and sftp and added them also to 'Server URL'.
Should we just delete http and https from the code?

MichaIng commented:
@DavHau 
If you find time, could you rebase onto current dev? Generally WebDAV support makes much sense. When you do, please leave out any changes which are not directly related/required for the additional upload service(s) to work. Good fine about the unused `http|https`, but please do this as an own PR, so we can test this individually.

Also I agree, if this requires an additional dependency, it should be probably made optional. Either the existence of the module/function/method could be tested to display the option or not, or it could be made a config option, like SMB/CIFS is, with the additional dependency explained in the config sample.

DavHau commented:
I currently do not have the capacity to work on this. Also I'm not a user of motioneye anymore. Feel free to fork/copy and continue the work.

MichaIng commented:
Okay, fair enough.

@jmichault @kleini @bob-lee @cclauss shall we keep this PR open, in the hope someone implements WebDAV support based on it, or shall we close it?

MichaIng commented:
Cross linking a similar PR with only Nextcloud upload added: #1743

MichaIng commented:
All done now in #1743, marking this as closed.
